### ISSUE TITLE
Hang, when reading VDS if mismatch between source and virtualsource shapes

### ISSUE THREAD

**Issue by mrosanes** [h5py/h5py#1218]
-    Operating System: Debian9
-    Python version: 2.7.13
-    Where Python was acquired: conda environment
-    h5py version: 2.9.0
-    HDF5 version: 1.10.4

If a mismatch exist between source and virtualsource shapes, at the VDS creation; and then we try to read the virtual dataset, h5py is hanging. It should be good if an exception was reported when trying to read a VDS that was incorrectly created, with source shape and virtualsource shape not matching.

The hang can be reproduced with the following python script:
[source_smallerthan_vs.txt](https://github.com/h5py/h5py/files/3160891/source_smallerthan_vs.txt)

This script creates a VDS, with source datasets of shape 2x2; and it is indicated that the virtual datasets would have a shape of 3x3. 

Then we try to read a slice from the VDS, and it hangs:
img = fd['dataset'][0]

This is also observed with programs that try to read the same VDS (e.g. silx).

**Comment by takluyver** [comment]
When you read a dataset, h5py doesn't know whether it's a virtual dataset or not: it just calls the HDF5 C functions to read the dataset. So it's possible that this is actually a problem in libhdf5.

I wonder if it would make sense for h5py to offer some means of checking a virtual dataset. We have something like this at work because when a virtual dataset points to data you can't access, reading it gives you the fill value instead of any error.

-----

Inlining the attached code to see it easier:

```python
import h5py
import numpy as np

dataset_name = 'dataset'

def createSourceFiles(n_source_files):
    ret = []
    # Create source files (1.h5, 2.h5...)
    for n in range(1, n_source_files+1):
        name = '{}.h5'.format(n)
        ret.append(name)
        with h5py.File(name, 'w') as f:
            d = f.create_dataset(dataset_name, (2, 2), 'i4')
            d[:] = np.random.randint(n_source_files, size=(2, 2))
    return ret


def createVDS(sourcefiles):
    
    with h5py.File("vds.h5", mode='w') as fd:
        # Assemble virtual dataset
        num_files = len(sourcefiles)
        layout = h5py.VirtualLayout(shape=(num_files, 3, 3), dtype='i4')
        
        for i, fn in enumerate(sourcefiles):
            vsource = h5py.VirtualSource(fn, dataset_name, shape=(3, 3))
            layout[i] = vsource
            
        vdataset = fd.create_virtual_dataset("dataset", layout,
                                             fillvalue=-np.inf)      

def show_vds_file():
    with h5py.File("vds.h5", mode='r') as fd:
        img = fd['dataset'][0]
        print(img)


sourcefiles = createSourceFiles(3)
createVDS(sourcefiles)
show_vds_file()
```

**Comment by mrosanes** [comment]
> When you read a dataset, h5py doesn't know whether it's a virtual dataset or not: it just calls the HDF5 C functions to read the dataset. So it's possible that this is actually a problem in libhdf5.

Should we report this to libhdf5, then?


> I wonder if it would make sense for h5py to offer some means of checking a virtual dataset. 

+1

> We have something like this at work because when a virtual dataset points to data you can't access, reading it gives you the fill value instead of any error.

We also noticed this and we think that it should be considered a bug. If you already have some shareable implementation, we would be very interested in using it.

**Comment by takluyver** [comment]
> Should we report this to libhdf5, then?

Possibly. It would be ideal if someone could replicate it using the C API first, to show conclusively that it's not something h5py is doing wrong. But it's not obvious to me how h5py's code could result in a hang.

> We also noticed this and we think that it should be considered a bug.

I have a suspicion it might be baked into the design of virtual datasets, unfortunately. But if you want to post on the HDF forum about it, I'd be interested to see if we can get any traction.

I posted about a specific flavour of this issue, but I haven't heard anything yet: https://forum.hdfgroup.org/t/virtual-dataset-in-read-write-file-missing-data-from-read-only-file/5647

> If you already have some shareable implementation, we would be very interested in using it.

It's in an internal repository at the moment, but I think it should be fine to move it to a public repository and make it a standalone tool. There's not much to it, though.

**Comment by mrosanes** [comment]
> It's in an internal repository at the moment, but I think it should be fine to move it to a public repository and make it a standalone tool. There's not much to it, though.

This would be great, even if there's not much to it; many thanks.

> I posted about a specific flavour of this issue, but I haven't heard anything yet: https://forum.hdfgroup.org/t/virtual-dataset-in-read-write-file-missing-data-from-read-only-file/5647

I tried to reproduce this issue (following your notebook), with the conda environment described at the top of this ticket; and in my case I can read correctly the vds using 'r+' mode. Both rows of the virtual dataset are read correctly in 'r+' mode also.

**Comment by takluyver** [comment]
Thanks, that's strange - I've got the same versions of h5py and HDF5 as you. I'm on Fedora, but I don't think that should make a difference.

Can you try opening the read-only source file directly? I get an error from this:

```python
s2 = h5py.File('src2.h5', 'r+')
```

**Comment by mrosanes** [comment]
Indeed it is strange; I also saw that we are using the same versions of hdf5 (1.10.4) and h5py (2.9.0).

(I use python 2.7.13 and numpy 1.16.3)

Tried, I can do this correctly; both of them:

<pre>s2 = h5py.File('src2.h5', 'r+')</pre>

<pre>s1 = h5py.File('src1.h5', 'r+')</pre>

**Comment by takluyver** [comment]
If you then try to modify s2 (e.g. add an extra dataset to it), what happens? Is it possible that for some reason it's not actually read-only?

**Comment by mrosanes** [comment]
You have reason; sorry, my mistake; I was not removing correctly the write permissions from src2.h5.

**Comment by takluyver** [comment]
OK, no worries. So when you do remove the write permissions, you can replicate what I saw?

**Comment by mrosanes** [comment]
Yes, I have exactly the same behavior that you posted in the hdf5 forum.

**Comment by takluyver** [comment]
Great, thanks for checking it :-)

**Comment by mrosanes** [comment]
you're welcome :)

**Comment by takluyver** [comment]
I've now made the virtual-dataset checking script into a little standalone tool and published it here: https://github.com/European-XFEL/hdf5-vds-check

**Comment by mrosanes** [comment]
Many thanks Thomas, it will be useful for us.
